Automattic/wp-calypso
# Payments: Move lib/stripe to calypso-stripe package

#### Changes proposed in this Pull Request

This moves the `lib/stripe` directory (which is used to access [Stripe Elements](https://stripe.com/payments/elements) for our credit card form fields) to a new npm package in the monorepo called `@automattic/calypso-stripe`.

In the process of this move, I converted the functions in that file to TypeScript (it's easiest to review that change by looking at the commits in this PR one-by-one). While doing this, I tried not to modify the function signatures at all, but a few changes were necessary, which I will list below.

- `StripeHookProvider` now has `fetchStripeConfiguration` as a required prop. Previously this was optional, and the default was to use the bound wpcom.js functions. However, the authentication needed for that library is entangled deeply with calypso, making it much easier to inject it from the consumer components rather than try to import it wholesale. All implementations have been updated to have this prop.
- `StripeHookProvider` now has an optional `locale` prop. In https://github.com/Automattic/wp-calypso/pull/43026, we added a mechanism to pass the user's locale to Stripe.js which it will use to localize any error messages it creates. However, that data comes from the calypso Redux store and is therefore tightly coupled to the calypso package. It is easier to inject the string from the consumer components rather than try to pull it directly within the new package. All implementations (other than tests, and [unused components](https://github.com/Automattic/wp-calypso/pull/46920)) have been updated to have this prop.

This also implements the changes from https://github.com/Automattic/wp-calypso/pull/46953 and https://github.com/Automattic/wp-calypso/pull/47002

#### Testing instructions

Because of the move itself, as well as the modifications to `StripeHookProvider` above, this touches quite a few places in calypso. They should not be affected, but it's worth testing all of them to be sure. For the most part, if you visit these pages and the credit card form loads without any errors (either in the UI or in the console), then this should be fine.

<img width="937" alt="Screen Shot 2020-11-04 at 11 12 28 AM" src="https://user-images.githubusercontent.com/2036909/98136207-aeb1a880-1e8e-11eb-93be-70d64deec34b.png">

- Checkout (visit `/plans`, click to add a plan and you will be redirected to `/checkout`).
- Add an account-level card (visit `/me/purchases/add-credit-card`).
- Add a site-level card (visit `/purchases/add-credit-card`).
- Update a card (visit `/purchases`, click on a purchase _that was not made with credits_, and then click "Change payment method").
- Add a card to WooCommerce label services (see the testing instructions in https://github.com/Automattic/wp-calypso/pull/35390).
- Checkout from the block editor (see the testing instructions in https://github.com/Automattic/wp-calypso/pull/46860).

(This also touches old checkout for the sake of completeness, but there's no need to test that because it is unused and that code will soon be removed.)
